### PR TITLE
[FIX] Resource: getStream for empty string

### PR DESCRIPTION
**Pull request by tobiasso85**
If a resource does not have any content (empty string), the stream still can be retrieved.
Before the error `Resource ${this._path} has no content`
was thrown for the empty string.

Add additional test "getStream for empty file: correctly retrieved" which uses a buffer for an empty file for completeness.

### CODE REVIEW

File: test/lib/adapters/FileSystem_read.js [hunk 1]
```diff
@@ -94,7 +94,7 @@ test("glob resources from application.a with directory exclude", async (t) => {
 	});
 
 	await readerWriter.byGlob("/!(pony,unicorn)/**").then(function(resources) {
-		t.deepEqual(resources.length, 2, "Found exactly two resource");
+		t.deepEqual(resources.length, 3, "Found exactly two resource");
```
Comment (RandomByte) [inline test/lib/adapters/FileSystem_read.js:97]:
```suggestion
		t.deepEqual(resources.length, 3, "Found exactly three resource");
```

File: test/lib/Resource.js [hunk 2]
```diff
@@ -160,6 +185,12 @@ test("Resource: clone resource with stream", (t) => {
 	});
 });
 
+test("getStream for empty file: correctly retrieved", async (t) => {
```
Comment (RandomByte) [inline test/lib/Resource.js:188]: This test is unrelated to your fix. It also passes without it. Why is it necessary? The commit message does not explain that.

**Comment by RandomByte** [comment]
Also, this should be a `[FIX]` so that it is visible in the changelog

File: test/lib/Resource.js (hunk 2 above)
Comment (tobiasso85) [reply test/lib/Resource.js:188]: added it for completeness, there was no test for a stream of an empty file. Will add it to the commit message.

File: test/lib/adapters/FileSystem_read.js (hunk 1 above)
Comment (tobiasso85) [reply test/lib/adapters/FileSystem_read.js]: done

Comment (tobiasso85) [reply test/lib/adapters/FileSystem_read.js]: done

File: test/lib/Resource.js (hunk 2 above)
Comment (tobiasso85) [reply test/lib/Resource.js:188]: will update message

Comment (RandomByte) [reply test/lib/Resource.js:188]: Well I'm just wondering because half of the changes done in this PR are related to this test. And I'm not sure whether it tests any of our code.

In this test case, you pass in a stream (using the `createStream` callback) and retrieve it using `getStream`. The Resource implementation never handles the streams content. So it doesn't matter whether it's empty or not. Maybe I'm missing something but to me this test seems pretty useless?

Comment (tobiasso85) [reply test/lib/Resource.js]: removed it

File: test/lib/Resource.js [hunk 3]
```diff
@@ -4,10 +4,10 @@ const fs = require("fs");
 const path = require("path");
 const Resource = require("../../lib/Resource");
 
-function createBasicResource() {
-	const fsPath = path.join("test", "fixtures", "application.a", "webapp", "index.html");
+function createBasicResource(fileName = "index.html") {
```
Comment (RandomByte) [inline test/lib/Resource.js:7]: This change is not necessary anymore

File: test/lib/adapters/FileSystem_read.js [hunk 4]
```diff
@@ -94,7 +94,7 @@ test("glob resources from application.a with directory exclude", async (t) => {
 	});
 
 	await readerWriter.byGlob("/!(pony,unicorn)/**").then(function(resources) {
-		t.deepEqual(resources.length, 2, "Found exactly two resource");
+		t.deepEqual(resources.length, 3, "Found exactly three resource");
```
Comment (RandomByte) [inline test/lib/adapters/FileSystem_read.js:97]: test/fixtures/application.a/webapp/empty.js can be removed again and this change reverted

File: test/lib/Resource.js (hunk 3 above)
Comment (tobiasso85) [reply test/lib/Resource.js]: done

File: test/lib/adapters/FileSystem_read.js (hunk 4 above)
Comment (tobiasso85) [reply test/lib/adapters/FileSystem_read.js]: done

File: lib/Resource.js [hunk 5]
```diff
@@ -73,7 +73,7 @@ class Resource {
 		this._createStream = createStream || null;
 		this._stream = stream || null;
 		this._buffer = buffer || null;
-		if (string) {
+		if (typeof string === "string") {
```
Comment (RandomByte) [inline lib/Resource.js:76]: Should we also check for `string instanceof String` to catch cases where a `new String("bla")` object is passed?

Comment (tobiasso85) [reply lib/Resource.js:76]: I would either use then both in combination:
typeof and instanceof

```js
let simpleStr = 'This is a simple string'
let myString  = new String()
let newStr    = new String('String created with constructor')

simpleStr instanceof String  // returns false, string literal is not an object
myString  instanceof String  // returns true
newStr    instanceof String  // returns true
(newStr + "a") instanceof String  // returns false
newStr.toString() instanceof String  // returns false
```
source + modified:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/instanceof

or leave typeof because that is rare that a string is constructed using the constructor.

Comment (RandomByte) [reply lib/Resource.js:76]: Yes, I would add that as an additional check as well. Just like here: https://stackoverflow.com/a/9436948

> or leave typeof because that is rare that a string is constructed using the constructor with new.
Yes it's rare, but if somebody does it we would still create the resource but without any content. That might cause issues that are not easy to understand. We should add this check to prevent this from happening.

Comment (tobiasso85) [reply lib/Resource.js]: kk, done

File: test/lib/Resource.js [hunk 6]
```diff
@@ -62,20 +83,38 @@ test("Resource: getStream", async (t) => {
...
-			resolve(streamedResult);
-		});
-	}).then((result) => {
+	return readStream(resource.getStream()).then((result) => {
 		t.is(result, "Content", "Stream has been read correctly");
 	});
 });
 
+test("Resource: getStream for empty string", async (t) => {
+	t.plan(1);
```
Comment (RandomByte) [inline test/lib/Resource.js:92]: Not necessary but doesn't hurt
